File: src/main/java/me/syari/ss/discord/internal/requests/CallbackContext.java
package me.syari.ss.discord.internal.requests;

public class CallbackContext implements AutoCloseable {
    private static final ThreadLocal<Boolean> callback = ThreadLocal.withInitial(() -> false);
    private static final CallbackContext instance = new CallbackContext();

    public static CallbackContext getInstance() {
        startCallback();
        return instance;
    }

    static boolean isCallbackContext() {
        return callback.get();
    }

    private static void startCallback() {
        callback.set(true);
    }

    @Override
    public void close() {
        callback.set(false);
    }
}

File: src/main/kotlin/me/syari/ss/discord/requests/CallbackContext.kt
package me.syari.ss.discord.requests

internal class CallbackContext private constructor(): AutoCloseable {
    override fun close() {
        callback.set(false)
    }

    companion object {
        private val callback = ThreadLocal.withInitial { false }

        val instance = CallbackContext()
            get() {
                startCallback()
                return field
            }

        val isCallbackContext: Boolean
            get() = callback.get()

        private fun startCallback() {
            callback.set(true)
        }
    }
}